### PR TITLE
Persisting hashrange in lease table

### PR DESCRIPTION
**Pull request by ashwing**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import lombok.experimental.Accessors;
+import org.apache.commons.lang3.Validate;
+import software.amazon.awssdk.services.kinesis.model.HashKeyRange;
+
+@Data
+@Accessors(fluent = true)
+/**
+ * Lease POJO to hold the starting hashkey range and ending hashkey range of kinesis shards.
+ */
+public class HashKeyRangeForLease {
```
Comment (joshua-kim) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java:30]: Is there a reason we need to maintain the range? Can we just have two separate DDB fields, StartingHashKey and EndingHashKey so we can avoid doing this parsing altogether?

Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java:30]: Just wanted to make both the updates atomic.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseSerializer.java [hunk 2]
```diff
@@ -54,6 +55,7 @@
     private static final String PENDING_CHECKPOINT_STATE_KEY = "pendingCheckpointState";
     private static final String PARENT_SHARD_ID_KEY = "parentShardId";
     private static final String CHILD_SHARD_IDS_KEY = "childShardIds";
+    private static final String HASH_KEY_RANGE = "hashKeyRange";
```
Comment (joshua-kim) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseSerializer.java]: Refer to earlier comment

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/Lease.java [hunk 3]
```diff
@@ -274,6 +281,18 @@ public void childShardIds(@NonNull final Collection<String> childShardIds) {
         this.childShardIds.addAll(childShardIds);
     }
 
+    /**
+     * Set the hash range key for this shard.
+     * @param hashKeyRangeForLease
+     */
+    public void hashKeyRange(final HashKeyRangeForLease hashKeyRangeForLease) {
+        if(this.hashKeyRangeForLease == null) {
```
Comment (ychunxue) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/Lease.java]: Seems in the constructor, we are doing null check. But in the setter we are not. Shall we keep this consistent?

**Review by ychunxue — APPROVED** [review]
LGTM except minor comment.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     * Deserialize from serialized hashKeyRange string from external storage.
+     * @param hashKeyRange
+     * @return HashKeyRangeForLease
+     */
+    public static HashKeyRangeForLease deserialize(String hashKeyRange) {
+        final String[] hashKeyTokens = hashKeyRange.split(DELIM);
+        Validate.isTrue(hashKeyTokens.length == 2, "HashKeyRange should have exactly two tokens");
+        // Assuming that startingHashKey and endingHashRange are not same.
+        Validate.isTrue(!hashKeyTokens[0].equals(hashKeyTokens[1]), "StartingHashKey and EndingHashKey should not be same");
+        Validate.isTrue(hashKeyTokens[0].compareTo(hashKeyTokens[1]) < 0, "StartingHashKey must be less than EndingHashKey");
```
Comment (satselva) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java]: As disccussed this cannot be string

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+@Accessors(fluent = true)
+/**
+ * Lease POJO to hold the starting hashkey range and ending hashkey range of kinesis shards.
+ */
+public class HashKeyRangeForLease {
+
+    private static final String DELIM = ":";
+
+    private final String startingHashKey;
+    private final String endingHashKey;
```
Comment (satselva) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java]: Change it to BigInt as in ShardInfo

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseSerializer.java [hunk 6]
```diff
@@ -85,6 +87,10 @@
             result.put(PENDING_CHECKPOINT_STATE_KEY, DynamoUtils.createAttributeValue(lease.checkpoint().subSequenceNumber()));
         }
 
+        if(lease.hashKeyRangeForLease() != null) {
```
Comment (satselva) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseSerializer.java:91]: How does the existing Leases in LeaseTable will have the hash ranges updated for the lease ? Are we updating all the fields in Lease on every checkpoint ?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java (hunk 1 above)
Comment (satselva) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java:30]: The updates to DDB are atomic whether you do on single field or two fields in DDB. I would prefer having it as two fields in DDB and avoid this parsing logic.  (Having it as range stops us to create indexes if needed in future)

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import lombok.Data;
+import lombok.experimental.Accessors;
+import org.apache.commons.lang3.Validate;
+import software.amazon.awssdk.services.kinesis.model.HashKeyRange;
+
+@Data
+@Accessors(fluent = true)
+/**
+ * Lease POJO to hold the starting hashkey range and ending hashkey range of kinesis shards.
+ */
```
Comment (satselva) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java:29]: We can avoid this class, if we have HashRangeStart and End as two fields in DDB

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewerTest.java [hunk 8]
```diff
@@ -55,7 +56,8 @@
     private LeaseRefresher leaseRefresher;
 
     private static Lease newLease(String leaseKey) {
```
Comment (satselva) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewerTest.java:58]: Is there any validation in the test cases that Lease contains the hash range available in the Shard information ?  CAn we add any assert statements for it

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java (hunk 1 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java:30]: Not the DDB updates. I wanted this to be updated and maintained in a pair across the KCL code.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/Lease.java (hunk 3 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/Lease.java]: The one in constructor is for optimization and not for safety. Keeping it consistent.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java (hunk 4 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java]: ya updating in new revision

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java (hunk 5 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java]: I don't see a purpose of changing this to BigInteger as of now. But might be helpful during hashrangecompletion validation.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseSerializer.java (hunk 6 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseSerializer.java:91]: The existing leases will have them updated as part of periodic shard syncer, when it finds missing hashrange fields during hashrange completion validation.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java (hunk 1 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/HashKeyRangeForLease.java:30]: I will check if we can have different fields. But I dont see if we might introduce indices to KCL lease table.

**Review by joshua-kim — APPROVED** [review]
LGTM

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewerTest.java [hunk 9]
```diff
@@ -55,7 +56,8 @@
...
-        return new Lease(leaseKey, "LeaseOwner", 0L, UUID.randomUUID(), System.nanoTime(), null, null, null, new HashSet<>(), new HashSet<>(), null);
+        return new Lease(leaseKey, "LeaseOwner", 0L, UUID.randomUUID(), System.nanoTime(), null, null, null,
```
Comment (satselva) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewerTest.java:59]: Any validation for the hash range in the lease ?

**Review by satselva — APPROVED** [review]
Looks good to me !!

Minor comment, we could add any validation on the hash range for the lease in test cases

**Pull request merged**